pgpubsub/channel.py

```diff
@@ -0,0 +1,170 @@
+import datetime
+import inspect
+import json
+from abc import abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass
+from pydoc import locate
+
+import django
+from django.apps import apps
+from django.db import connection
+
+registry = defaultdict(list)
+
+
+@dataclass
+class _Channel:
+
+    def __post_init__(self):
+        self.callbacks = []
+
+    @classmethod
+    def name(cls):
+        # TODO: check channel name limits
+        module_name = inspect.getmodule(cls).__name__
+        name =  f'{module_name}_{cls.__name__}'
+        # TODO: pg_notify accepts ., but LISTEN does not.
+        # What's the best way to build a channel name?
+        name = name.replace('.', '_')
+        return name.lower()
+
+    @classmethod
+    def get(cls, name):
+        for channel_cls, callbacks in registry.items():
+            if channel_cls.name() == name:
+                return channel_cls, callbacks
+
+    @classmethod
+    def register(cls, callback):
+        registry[cls].append(callback)
+
+    @classmethod
+    @abstractmethod
+    def deserialize(cls, payload):
+        return
+
+    @classmethod
+    def build_from_payload(cls, notification_payload, callbacks):
+        deserialized = cls.deserialize(notification_payload)
+        channel = cls(**deserialized)
+        channel.callbacks.extend(callbacks)
+        return channel
+
+    @property
+    def signature(self):
+        return {k: v for k, v in self.__dict__.items()
+                if k in self.__dataclass_fields__}
+
+    def execute_callbacks(self):
+        for callback in self.callbacks:
+            callback(**self.signature)
+
+
+@dataclass
+class Channel(_Channel):
+
+    @classmethod
+    def deserialize(cls, payload):
```

Comment:
Im assuming you're doing a nested deserialization of dataclasses here.

It might be worth looking into https://github.com/monomonedula/nvelope for code that already does this.

I looked into it briefly and noticed it did hierarchical dataclass serialization/deserialization. Might eliminate this code.

I only briefly looked into the lib though

Reply:
Oh cool thanks! I did mention in my original notes that I think this part is likely re-inventing the wheel, but could actually not find anything which actually does this already. I'll definitely check this out. 
Right now the serialization logic only works for (iterables of) primitives + naive date objects, so maybe this library will also provide more.